addons/website_event_crm/models/event_lead_rule.py

```diff
@@ -0,0 +1,36 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import models
+from odoo.http import request
+
+from odoo.addons.website.models import ir_http
+
+
+class EventLeadRule(models.Model):
+
+    _inherit = "event.lead.rule"
+
+    def _run_on_registrations(self, registrations):
```

Comment:
Three main comments:

1. Performance

This will create leads in super, and then update their values here, effectively making multiple queries (insert+update) to set that information.
It seems like gathering values from the registration is the purpose of "_get_lead_values" on the event.registration model, any reason to not put the logic there in a website_event_crm override?

2. Complexity

Do we actually have use cases where the language from the visitor is NOT the one we want?
Main use case should be: I come on the website, change the language to what I want (should be propagated to visitor record) and then book my tickets to the event.
So we could just: set visitor from registration record + set language from that visitor if it exists

Maybe there are corner cases to this, can you elaborate if so? If very niche we may ignore to simplify.
(Or maybe it does not work at all, in which case I'm curious as to why).

3. Test

The lead generation based on rules is a bit complex, so we have a complete test suite in "test_event_crm_flow.py". It would be nice to add a test case for the visitor/lang combo inside "website_event_crm" as well.
(No need to re-test complex cases, just that we correctly propagate).

Reply:
Hello @awa-odoo,

To answer your points in order:

1) seems like a better solution to limit queries indeed

2) the reason we prioritize using `request.lang` instead of `visitor.lang_id` is actually that, counter-intuitively, the visitor's language is never updated. Meaning that if the visitor visits the website in english and then switches to french, their language will always be english. This is why, to get a more accurate representation of the visitor's actual language, we decided to get the language from the request instead

3) ok for a test

Reply:
Hello @adrienmilis 

I re-discussed this with Luc and we will use the visitor language after all, even if not optimal.

The main reason is that the "request" object is very volatile.
Imagine another button for the Odoo admin that runs "_run_on_registrations" on, say a set of selected records, it would use the language of that admin user instead, which is not what we want.
Basically, we want this logic to use proper records and not be context-dependent.

If you have further questions, feel free to hit me here or on Discord :slightly_smiling_face: 

Cheers!

Reply:
Hello @awa-odoo,

I'm in the process of simplifying. The case you described wouldn't happen, since before using `request.lang`, we discriminate between front-end and back-end.

`request.lang` is only used whenever `_run_on_registrations` is triggered after a call from the front-end, i.e. when a visitor has registered through the website.

I don't see a case where the function is called from the front-end and we don't want the request's lang, since it can only mean a visitor has registered to an event. However, if you still think this is dangerous, I can scrap it

```python
def _get_lead_values(self, rule):
    lead_values = super()._get_lead_values()

    # `get_request_website` returns the website record if from frontend or False if from backend
    lead_values.update({
        'visitor_ids': self.visitor_id,
        'lang_id': request.lang if ir_http.get_request_website() else self.visitor_id.lang_id
    })
    return lead_values
```

Thanks for the review!